src/main/cpp/src/cast_string_to_timestamp.cu

```diff
@@ -0,0 +1,876 @@
+
+namespace {
+
+enum class RESULT_TYPE : uint8_t {
+  // Parse success
+  SUCCESS = 0,
+
+  // Parse failed, has invalid format
+  // Throw exception when it's Ansi mode, or return null otherwise.
+  INVALID = 1
+};
+
+/**
+ * If the timestamp string only contains the time part, TS_TYPE is JUST_TIME.
+ * E.g.:
+ *   For timestamp string T18:01:01 : TS_TYPE = JUST_TIME
+ *   For timestamp string 2020-01-01T12:00:00 : TS_TYPE = NOT_JUST_TIME
+ */
+enum class TS_TYPE : uint8_t { NOT_JUST_TIME = 0, JUST_TIME = 1 };
+
+/**
+ * Represents local date time in a timezone.
+ * Spark stores timestamp into Long in microseconds.
+ * A Long is able to represent a timestamp within [+-]200 thousand years.
+ * So year is 4-6 digits, formula is: Long.MaxValue/microseconds_per_year.
+ */
+
+enum class TZ_TYPE : uint8_t {
+
+  // Not specified timezone in the string, indicate to use the default timezone.
+  NOT_SPECIFIED = 0,
+
+  // Fixed offset timezone
+  // String starts with UT/GMT/UTC/[+-], and it's valid.
+  // E.g: +08:00, +08, +1:02:30, -010203, GMT+8, UTC+8:00, UT+8
+  // E.g: +01:2:03
+  FIXED_TZ = 1,
+
+  // Not FIXED_TZ, it's a valid timezone string.
+  // E.g.: java.time.ZoneId.SHORT_IDS: CTT
+  // E.g.: Region-based timezone: America/Los_Angeles
+  OTHER_TZ = 2,
+
+  // Invalid timezone.
+  // String starts with UT/GMT/UTC/[+-], but it's invalid.
+  // E.g: UTC+19:00, GMT+19:00, max offset is 18 hours
+  // E.g: GMT+01:2:03, +01:2:03, special case
+  // E.g: non-exist-timezone
+  INVALID_TZ = 3
+};
+
+/**
+ * Represents a timezone.
+ */
+struct time_zone {
+  TZ_TYPE type;
+
+  // for fixed timezone
+  int fixed_offset;
+
+  // for java.time.ZoneId.SHORT_IDS and Region-based zone IDs
+  // the tz offset to the input string
+  int tz_pos_in_string;
+
+  // for java.time.ZoneId.SHORT_IDS and Region-based zone IDs
+  // the tz end offset to the input string
+  int tz_end_pos_in_string;
+
+  __device__ int tz_len() const { return tz_end_pos_in_string - tz_pos_in_string; }
+
+  __device__ time_zone()
+    : type(TZ_TYPE::NOT_SPECIFIED), fixed_offset(0), tz_pos_in_string(0), tz_end_pos_in_string(0)
+  {
+  }
+  __device__ time_zone(TZ_TYPE t, int offset, int tz_pos, int tz_end_pos)
+    : type(t), fixed_offset(offset), tz_pos_in_string(tz_pos), tz_end_pos_in_string(tz_end_pos)
+  {
+  }
+};
+
+__device__ time_zone make_fixed_tz(int offset)
+{
+  return time_zone(TZ_TYPE::FIXED_TZ, offset, 0, 0);
+}
+
+__device__ time_zone make_invalid_tz() { return time_zone(TZ_TYPE::INVALID_TZ, 0, 0, 0); }
+
+__device__ time_zone make_other_tz(int tz_pos, int tz_end_pos)
+{
+  return time_zone(TZ_TYPE::OTHER_TZ, 0, tz_pos, tz_end_pos);
+}
+
+/**
+ * Is white space, consistent with Spark UTF8String.trimAll for char input
+ */
+__device__ bool is_whitespace(char const c)
+{
+  // Keep consistent with Java: Character.isWhitespace(c) ||
+  // Character.isISOControl(c),
+  // 0-31 is control characters, 32 is space, 127 is delete
+  return (c >= 0 && c <= 32) || c == 127;
+}
```

Comment:
This will lead to the same error as in https://github.com/NVIDIA/spark-rapids-jni/issues/3227.

Maybe copy this? https://github.com/NVIDIA/spark-rapids-jni/pull/3250/files#diff-3f32312ca5b95c20e7c0aeee7fd0d6271e17474fc013e97ffcd057e6f410918bR48-R56

However, why do we care about code 127 here?

Reply:
UTF8String.trimAll treat ascii 127 as space.
`Character.isISOControl(127) = true`

Details:
https://github.com/apache/spark/blob/v3.5.0/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L412
```scala
def parseTimestampString
...
  val bytes = s.trimAll().getBytes
```

https://github.com/apache/spark/blob/v3.5.0/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
```java
  public UTF8String trimAll() {
    ... 
    while (s < this.numBytes && isWhitespaceOrISOControl(getByte(s))) s++;
  }

  private boolean isWhitespaceOrISOControl(int codePoint) {
    return Character.isWhitespace(codePoint) || Character.isISOControl(codePoint);
  }
```